docs: add remaining MVP docs content

yusefnapora opened the pull request:
Adds "How to retrieve NFT data from IPFS" guide to the new docs branch.

There are still a few things to do,  but I figured it's best to kick this off my laptop as a draft instead of holding onto it until it's perfect or whatever.

To do:

- [x] add local IPFS examples
- [x] show how to export an NFT as a CAR

Eventually it would be sweet to get IPFS desktop into the mix, but I hit [some issues](https://github.com/ipfs/ipfs-desktop/issues/1962) with dag-cbor support that make it kind of awkward at the moment.

dchoi27 commented:
Looks good! 

The flat structure of the subsections is a bit confusing - maybe we say "Retrieval option:" before the gateways and using your computer sections?

Also might suggest adding some thoughts saying that, in the future, wallets and other NFT tools being able to archive data automatically for you, and that users should ask their favorite ones to add this feature!

dashcraft commented:
looks like this needs a rebase/merge @yusefnapora

yusefnapora commented:
Good call @dashcraft :)

Everything should be up to date now, and I added a simple JS example to the quickstart.

I also added a conclusion to the "mint with custom metadata" how-to. I kind of punted on the actual minting part, since there's so many potential ways to do it, and if you already know you need custom metadata, chances are you know what to do better than we could guess... So I just threw in some recommendations to use IPFS uris and linked out to the Flow tutorial on nft school.

I think this ticks all the MVP content boxes at last... so here's a mass summons to nft storage devs to ask for a tech-focused review to make I've got the details right: @alanshaw @Gozala @hugomrdias @vasco-santos @jchris @the-simian - basically, if anyone's got a minute to skim through the preview build from the latest commit & give it the old sanity check, that would be 💯

alanshaw commented:
I like where this is going 🚀. Here's my feedback:

* Any chance we can get syntax highlighting for the code snippets?

* The "Client" section in the nav is not good. Please remove or add proper content there.

* Line height on the page titles is too small:
    <img width="756" alt="Screenshot 2022-02-01 at 10 10 51" src="https://user-images.githubusercontent.com/152863/151950626-ac83d8a3-e451-4e28-b85a-11d1acd1d7f7.png">

    I added the tachyons "lh-title" class and it looks better, but it's being overriden by some other "nextra theme" class which I had to disable:
    <img width="1135" alt="Screenshot 2022-02-01 at 10 14 10" src="https://user-images.githubusercontent.com/152863/151951043-72ca504a-1c2e-4773-8d74-22696d79628c.png">

* Not sure on the solution for this...second hamburger not great also seems to be cut at around half the page height:
    <img width="411" alt="Screenshot 2022-02-01 at 10 22 49" src="https://user-images.githubusercontent.com/152863/151951680-f1fab628-9d58-47e6-a725-a5e348143824.png">

* Need more space around screenshots
    <img width="789" alt="Screenshot 2022-02-01 at 10 25 27" src="https://user-images.githubusercontent.com/152863/151952270-cbd2391e-f951-4438-b0eb-8e11123ca05a.png">

* Too much spacing between title and text in these callouts:
    <img width="786" alt="Screenshot 2022-02-01 at 10 25 41" src="https://user-images.githubusercontent.com/152863/151952388-7d25e799-3a22-4811-9530-7ac8c3596402.png">
 
* Internal page links don't work (in Firefox at least):
   <img width="207" alt="Screenshot 2022-02-01 at 10 27 38" src="https://user-images.githubusercontent.com/152863/151952614-3b8f6794-2eaa-48b3-a9be-b02368d57d33.png">
 
* Please can the blue for the hyperlinks be changed to something more in keeping with the theme?
* Can we lose the top level nav item (does not help) and move the rest in so they have more space
    <img width="254" alt="Screenshot 2022-02-01 at 10 32 05" src="https://user-images.githubusercontent.com/152863/151953049-750717f1-b1bf-4138-8b34-1efbbe26172f.png">
* More line height issues (double space in paragraphs and single space in bullets)
    <img width="814" alt="Screenshot 2022-02-01 at 10 33 50" src="https://user-images.githubusercontent.com/152863/151953233-d5a3319d-5aea-43b8-ae43-ef7b190ee464.png">

yusefnapora commented:
great feedback, @alanshaw, thanks.

I'm looking into getting code highlighting working now, but if @dashcraft has any ideas, please let me know :)

Also hoping to get the header links working, but failing that I can at least make sure those entries don't show in the side nav.

@dashcraft if you have any time to work on the CSS issues today or tomorrow, that would be sweet. Otherwise I'll see what I can grab this afternoon.

Gozala commented on packages/website/pages/docs/concepts/architecture-considerations.md:
```diff
@@ -42,5 +42,34 @@ Meanwhile, we're continuing to work on a flexible public key based authenticatio
+
+While some browsers like [Brave](https://brave.com) offer [native IPFS support][brave-ipfs], other browsers are restricted to HTTP and its related protocols, which makes engaging directly with the peer-to-peer protocol more difficult.
+
+As a platform operator, you can support all browsers by using an IPFS gateway to serve NFTs to your users via HTTP. Because NFT.Storage and other services following [best practices for NFT data on IPFS][ipfs-docs-nft-best-practices] use `ipfs://` URLs instead of `https://`, your web application should support re-writing IPFS URLs to target the gateway of your choice.
```
Maybe it's worth mentioning that as `ipfs://` gets more widely adopted rewriting would become less necessary and that is why you want to put `ipfs://` instead of `https://`, not sure if it can fit in few enough words to avoid distraction.

Maybe Opera is also worth mentioning

Gozala commented on packages/website/pages/docs/concepts/architecture-considerations.md:
```diff
@@ -42,5 +42,34 @@ Meanwhile, we're continuing to work on a flexible public key based authenticatio
+
+Because public gateways are a shared resource, you may prefer to [run your own dedicated gateway](https://blog.stacktical.com/ipfs/gateway/dapp/2019/09/21/ipfs-server-google-cloud-platform.html) to provide a more consistent user experience. Thanks to the flexibility of content addressing, you can direct your platforms users to your own gateway in your application layer. Anyone not using your plat ...
+
+Another potential optimization is to store redundant copies of your NFT data on a traditional storage service and deliver them via HTTP. This side-steps the IPFS retrieval entirely and may be more performant in some cases, but you'll need to do some bookkeeping to keep track of the HTTP locations of each piece of data. When pursuing this option, we strongly recommend exposing the original `ipfs:/ ...
```
I fail to understand how is this different from running your own gateway.

yusefnapora replied:
lol, good point! @dchoi27 what do you think? I think I added this from your initial outline, but maybe we don't want to recommend people rolling their own gateways.

dchoi27 replied:
i guess it depends on what is meant by "gateway" - one option refers to running your own IPFS nodes, the other is just using a mapping of CIDs to HTTP locations where you're not broadcasting what CIDs you have, but just know how to resolve HTTP requests for CIDs that you do have (kind of like Opensea does today)

I think it's fine to mention that folks can run their own gateways. We just have to provide a performant enough experience where it's not a good choice.

Gozala commented on packages/website/pages/docs/how-to/mint-custom-metadata.md:
```diff
@@ -167,7 +167,31 @@ See [this pull request](https://github.com/nftstorage/nft.storage/pull/991) for
+
+Although in many cases the most convenient way to [retrieve NFT data from IPFS][howto-retrieve] may involve using an IPFS HTTP gateway, you should avoid storing HTTP gateway links in a smart contract or other blockchain record.
+
+Instead, store the `ipfs://` URI, which doesn't depend on a single gateway provider. You can rewrite this URL into a gateway link at the "last mile" when displaying the NFT on the web. This ensures that the blockchain link is always valid as long as any IPFS peer is providing the data, and doesn't tie your NFT to any specific gateway host.
```
or even dns

Gozala commented on packages/website/pages/docs/how-to/retrieve.md:
```diff
@@ -2,16 +2,179 @@
+
+Here the `/amazing.gif` path points to an image file that's "inside" the IPFS directory with the CID `bafybeigvgzoolc3drupxhlevdp2ugqcrbcsqfmcek2zxiw5wctk3xjpjwy`. 
+
+URLs with the `ipfs://` prefix can be used directly with a browser that supports IPFS like [Brave](https://brave.com). Once you've [configured IPFS support][brave-ipfs], you can just paste an `ipfs://` URL into your address bar and see content in your browser.
```
Maybe we should turn those ipfs:// urls into links and instead say that if you are reading this in brave / opera or any other browser that supports ipfs:// URL you can just click those to see the content.

Gozala reviewed (approved):
💚 Awesome to see all this explained so well.

dashcraft commented:
> great feedback, @alanshaw, thanks.
> 
> I'm looking into getting code highlighting working now, but if @dashcraft has any ideas, please let me know :)
> 
> Also hoping to get the header links working, but failing that I can at least make sure those entries don't show in the side nav.
> 
> @dashcraft if you have any time to work on the CSS issues today or tomorrow, that would be sweet. Otherwise I'll see what I can grab this afternoon.

@yusefnapora 
For highlighting it uses prismjs so you have to use their syntax abbreviations, https://nextra.vercel.app/advanced/code-highlighting

I should be able to get to the CSS fixes this afternoon, yes. I think we can move docs along and handle the header links by forking and fixing their theme (and also customize it to what we'd like) in the near future

yusefnapora commented:
Hmm, well it looks like I'm still hitting [this issue with header links](https://github.com/shuding/nextra/issues/321) that @dashcraft ran into, but now it no longer affects production builds, just the dev server. So that's something :)

hugomrdias commented:
This looks awesome @yusefnapora 

Just a few UI nits

## The nav doesnt keep up with the header when you scroll
https://user-images.githubusercontent.com/314190/152144477-ccba31ae-e8f3-43e7-afcf-2982134248f5.mov

## In mobile theres a few overflows
<img width="404" alt="Screen Shot 2022-02-02 at 11 24 50" src="https://user-images.githubusercontent.com/314190/152144983-b2bc1797-ecf7-4161-9c7d-144f1bbac1a2.png">
<img width="405" alt="Screen Shot 2022-02-02 at 11 31 58" src="https://user-images.githubusercontent.com/314190/152145909-1020e7e8-f4dd-405d-8229-ab1e45d36176.png">

## Highlighting not working here
<img width="402" alt="Screen Shot 2022-02-02 at 11 27 06" src="https://user-images.githubusercontent.com/314190/152145233-04895f9d-9dfe-482e-a36d-0beaa4f5006c.png">

## refresh in docs render home page with broken links
https://user-images.githubusercontent.com/314190/152145727-a1cbd19f-c19b-48d9-9dab-e4c2f12bd3d7.mov


Any chance we can get typedoc to output markdown in embed here ?

Codigo-Fuentes commented:
<img width="796" alt="Screen Shot 2022-02-02 at 12 21 34 PM" src="https://user-images.githubusercontent.com/43380153/152185335-9373e4d3-83c2-4c29-ac1b-0ae4734d49e3.png">

**suggestion (non-blocking)**: Items in this list should be parallel--i.e., “creating, uploading, getting, using” or “create, upload, get, use”

yusefnapora commented:
Heads up to the team that I'm merging this into @dashcraft's integration branch so we can avoid conflicts as we go forward. Thanks for all the feedback! If there's more, add it to #1044 👍

JeffLowe commented:
Of note, I've captured all feedback above into flat list and will ticket out (so we can tackle critical launch-dependent items first, and then after launch tackle refinements).
